### PR TITLE
Differentiate CloudProviderName and KubeletCloudProviderName in templates

### PR DESCRIPTION
**Pull request by xmudrii**
**What this PR does / why we need it**:

We currently use the `CloudProviderName` field in templates for:

* the `--cloud-provider` flag in the Kubelet service
* running commands only for some provider(s)

However, `CloudProviderName` contains the value returned by `GetCloudConfig` function. This function returns the cloud provider name used by Kubelet, and only if the provider has an in-tree cloud provider implementation (in-tree CCM). In other words, `CloudProviderName` will be empty on providers such as Nutanix, DigitalOcean, Hetzner, etc...

We encountered the issue when we tried to install some packages only on Nutanix machines in #1161. Because Nutanix doesn't have in-tree CCM, `.CloudProviderName` is empty, and the Nutanix-specific logic is not executed.  Therefore, this PR is a prerequisite for #1161.

This issue is fixed by differentiating `CloudProviderName` and `KubeletCloudProviderName`:

* `CloudProviderName` is `providerConfig.CloudProvider`. It's used in templates when we want to run cloud-provider-specific logic (e.g. install some packages only for one provider)
* `KubeletCloudProviderName` is from `GetCloudConfig). It's passed to Kubelet via `--cloud-provider` flag

**Optional Release Note**:
```release-note
Differentiate CloudProviderName and KubeletCloudProviderName in templates
```

/assign @embik @moadqassem @ahmedwaleedmalik

### CODE REVIEW

File: pkg/userdata/flatcar/provider_test.go [hunk 1]
```diff
@@ -102,7 +101,7 @@ func (p *fakeCloudConfigProvider) GetCloudConfig(spec clusterv1alpha1.MachineSpe
 type userDataTestCase struct {
 	name                  string
 	spec                  clusterv1alpha1.MachineSpec
-	ccProvider            cloud.ConfigProvider
+	ccProvider            *fakeCloudConfigProvider
```
Comment (embik) [inline pkg/userdata/flatcar/provider_test.go:104]: I'm not sure I'm a fan of this. Doesn't this tell us that the `ConfigProvider` interface is missing something, like `func (p *ConfigProvider) Name() string`?

Comment (xmudrii) [reply pkg/userdata/flatcar/provider_test.go:104]: This is correct — having the `Name()` function in the `CloudProvider` interface would be helpful. However, that requires changing all cloud providers, and I don't have capacity for that right now. This is handled in a similar way for other providers where we just provide a string pointer and then based on it create `fakeCloudConfigProvider`.

Comment (xmudrii) [reply pkg/userdata/flatcar/provider_test.go]: Turns out this change was not needed at all. I reverted it.

**Review by ahmedwaleedmalik — APPROVED** [review]

**Pull request merged**